source/opt/aggressive_dead_code_elim_pass.cpp

```diff
@@ -290,39 +291,98 @@ Pass::Status AggressiveDCEPass::ProcessDebugInformation(
     std::list<BasicBlock*>& structured_order) {
   for (auto bi = structured_order.begin(); bi != structured_order.end(); bi++) {
     bool succeeded = (*bi)->WhileEachInst([this](Instruction* inst) {
-      // DebugDeclare is not dead. It must be converted to DebugValue in a
-      // later pass
-      if (inst->IsNonSemanticInstruction() &&
-          inst->GetShader100DebugOpcode() ==
+      if (!inst->IsNonSemanticInstruction()) return true;
+
+      if (inst->GetShader100DebugOpcode() ==
               NonSemanticShaderDebugInfo100DebugDeclare) {
-        AddToWorklist(inst);
-        return true;
-      }
+        uint32_t var_id =
+            inst->GetSingleWordInOperand(kDebugDeclareVariableInIdx);
+        auto var_def = get_def_use_mgr()->GetDef(var_id);
+
+        if (IsLive(inst)) return true;
+
+        if (IsLive(var_def)) {
+          AddToWorklist(inst);
+          return true;
+        }
 
-      // If the Value of a DebugValue is killed, set Value operand to Undef
-      if (inst->IsNonSemanticInstruction() &&
-          inst->GetShader100DebugOpcode() ==
+        // DebugDeclare Variable is not live. Find the value that was being
+        // stored to this variable. If it's live then create a new DebugValue
+        // with this value.  Otherwise let it die in peace.
+        uint32_t stored_value_id = 0;
+        get_def_use_mgr()->ForEachUser(
+            var_id, [&stored_value_id](Instruction* user) {
+              if (user->opcode() == spv::Op::OpStore) {
+                const uint32_t kStoreValueInIdx = 1;
+                stored_value_id =
+                    user->GetSingleWordInOperand(kStoreValueInIdx);
+              }
+            });
+
+        if (!IsLive(get_def_use_mgr()->GetDef(stored_value_id))) {
+          return true;
+        }
+
+        uint32_t value_id = stored_value_id;
+        if (value_id == 0) {
+          uint32_t type_id = var_def->type_id();
+          value_id = Type2Undef(type_id);
+          if (value_id == 0) return false;
+        }
+
+        Instruction* next_inst = inst->NextNode();
+        bool added = context()->get_debug_info_mgr()->AddDebugValueForVariable(
+            inst, var_id, value_id, inst);
```

Comment:
The debug value should be placed next to the store. See how it is done in `SSARewriter::ProcessStore`. If there is an initializer on the variable, then a debug value should be placed in the location of the debug declare. 

Your implementation will cause the debugger to think that the variable took on this value earlier than it actually does. Also, `value_id` might not dominate the location of the debug value. This could lead to invalid code.